cpp/tests/pagerank/pagerank_mg_test.cpp

```diff
@@ -0,0 +1,258 @@
+// setup and teardown steps as well as common utilities used by each E2E MG
+// test.  In this case, each test is identical except for the inputs and
+// expected outputs, so the entire test is defined in the run_test() method.
+//
+class Pagerank_E2E_MG_Testfixture_t : public ::testing::TestWithParam<Pagerank_Testparams_t> {
+ public:
+  Pagerank_E2E_MG_Testfixture_t() {}
+
+  // Run once for the entire fixture
+  //
+  // FIXME: consider a ::testing::Environment gtest instance instead, as done
+  // by cuML in test_opg_utils.h
+  static void SetUpTestCase()
+  {
+    MPI_TRY(MPI_Init(NULL, NULL));
+
+    int rank, size;
+    MPI_TRY(MPI_Comm_rank(MPI_COMM_WORLD, &rank));
+    MPI_TRY(MPI_Comm_size(MPI_COMM_WORLD, &size));
+
+    int nGpus;
+    CUDA_CHECK(cudaGetDeviceCount(&nGpus));
+
+    ASSERT(
+      nGpus >= size, "Number of GPUs are lesser than MPI ranks! ngpus=%d, nranks=%d", nGpus, size);
+
+    CUDA_CHECK(cudaSetDevice(rank));
+  }
+
+  static void TearDownTestCase() { MPI_TRY(MPI_Finalize()); }
+
+  // Run once for each test instance
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  // Return the results of running pagerank on a single GPU for the dataset in
+  // graph_file_path.
+  template <typename vertex_t, typename edge_t, typename weight_t, typename result_t>
+  std::vector<result_t> get_sg_results(raft::handle_t& handle,
+                                       const std::string& graph_file_path,
+                                       const result_t alpha,
+                                       const result_t epsilon)
+  {
+    auto graph =
+      cugraph::test::read_graph_from_matrix_market_file<vertex_t, edge_t, weight_t, true>(
+        handle, graph_file_path, true);  // FIXME: should use param.test_weighted instead of true
+
+    auto graph_view     = graph.view();
+    cudaStream_t stream = handle.get_stream();
+    rmm::device_uvector<result_t> d_pageranks(graph_view.get_number_of_vertices(), stream);
+
+    cugraph::experimental::pagerank(
+      handle,
+      graph_view,
+      static_cast<weight_t*>(nullptr),     // adj_matrix_row_out_weight_sums
+      static_cast<vertex_t*>(nullptr),     // personalization_vertices
+      static_cast<result_t*>(nullptr),     // personalization_values
+      static_cast<vertex_t>(0),            // personalization_vector_size
+      d_pageranks.begin(),                 // pageranks
+      alpha,                               // alpha (damping factor)
+      epsilon,                             // error tolerance for convergence
+      std::numeric_limits<size_t>::max(),  // max_iterations
+      false,                               // has_initial_guess
+      true);                               // do_expensive_check
+
+    std::vector<result_t> h_pageranks(graph_view.get_number_of_vertices());
+    raft::update_host(h_pageranks.data(), d_pageranks.data(), d_pageranks.size(), stream);
+
+    return std::move(h_pageranks);
```

Comment:
Doing a `move()` on a return does more harm than good in this case. Just do a simple `return h_pageranks;`. The compiler would perform a Return Value Optimization (RVO), a.k.a "copy elision" whereby it optimizes the allocation of the return value directly into the l-value that would hold the result of this call. Example:
```
auto return_v = get_sg_results(...);
```
The compiler would optimize the allocation of `h_pageranks` directly into `return_v`. However, if you invoke a `move()` you prevent that optimization from happening.

Reply:
Thanks, I will change that as suggested in the next commit to this PR.